data/transactions/logic/eval.go

```diff
@@ -2089,337 +2224,366 @@ func (cx *EvalContext) txnFieldToStack(txn *transactions.Transaction, fs txnFiel
+	// If last inner was a singleton, return it as a slice.
+	if gid.IsZero() {
+		return inners[len(inners)-1:]
+	}
+	// Look back for the first non-matching inner (by group) to find beginning
+	for i := len(inners) - 2; i >= 0; i-- {
+		if inners[i].Txn.Group != gid {
+			return inners[i+1:]
+		}
+	}
+	// All have the same (non-zero) group. Return all
+	return inners
+}
+
+func opGitxn(cx *EvalContext) {
+	gi := uint64(cx.program[cx.pc+1])
+	field := TxnField(cx.program[cx.pc+2])
+
+	sv, err := cx.opTxnImpl(gi, srcInnerGroup, field, 0, false)
+	if err != nil {
+		cx.err = err
 		return
 	}
 
-	itxn := &cx.InnerTxns[len(cx.InnerTxns)-1]
-	sv, err := cx.itxnFieldToStack(itxn, fs, arrayFieldIdx)
+	cx.stack = append(cx.stack, sv)
+}
+
+func opGitxna(cx *EvalContext) {
+	gi := uint64(cx.program[cx.pc+1])
+	field := TxnField(cx.program[cx.pc+2])
+	ai := uint64(cx.program[cx.pc+3])
+
+	sv, err := cx.opTxnImpl(gi, srcInnerGroup, field, ai, true)
 	if err != nil {
 		cx.err = err
 		return
 	}
+
 	cx.stack = append(cx.stack, sv)
 }
 
-func opGaidImpl(cx *EvalContext, groupIdx uint64, opName string) (sv stackValue, err error) {
-	if groupIdx >= uint64(len(cx.TxnGroup)) {
-		err = fmt.Errorf("%s lookup TxnGroup[%d] but it only has %d", opName, groupIdx, len(cx.TxnGroup))
+func opGitxnas(cx *EvalContext) {
+	last := len(cx.stack) - 1
+
+	gi := uint64(cx.program[cx.pc+1])
+	field := TxnField(cx.program[cx.pc+2])
+	ai := cx.stack[last].Uint
+
+	sv, err := cx.opTxnImpl(gi, srcInnerGroup, field, ai, true)
+	if err != nil {
+		cx.err = err
 		return
-	} else if groupIdx > cx.GroupIndex {
-		err = fmt.Errorf("%s can't get creatable ID of txn ahead of the current one (index %d) in the transaction group", opName, groupIdx)
+	}
+
+	cx.stack[last] = sv
+}
+
+func opGaidImpl(cx *EvalContext, gi int, opName string) (sv stackValue, err error) {
+	if gi >= len(cx.TxnGroup) {
+		err = fmt.Errorf("%s lookup TxnGroup[%d] but it only has %d", opName, gi, len(cx.TxnGroup))
 		return
-	} else if groupIdx == cx.GroupIndex {
+	} else if gi > cx.GroupIndex {
+		err = fmt.Errorf("%s can't get creatable ID of txn ahead of the current one (index %d) in the transaction group", opName, gi)
+		return
+	} else if gi == cx.GroupIndex {
 		err = fmt.Errorf("%s is only for accessing creatable IDs of previous txns, use `global CurrentApplicationID` instead to access the current app's creatable ID", opName)
 		return
-	} else if txn := cx.TxnGroup[groupIdx].Txn; !(txn.Type == protocol.ApplicationCallTx || txn.Type == protocol.AssetConfigTx) {
-		err = fmt.Errorf("can't use %s on txn that is not an app call nor an asset config txn with index %d", opName, groupIdx)
+	} else if txn := cx.TxnGroup[gi].Txn; !(txn.Type == protocol.ApplicationCallTx || txn.Type == protocol.AssetConfigTx) {
+		err = fmt.Errorf("can't use %s on txn that is not an app call nor an asset config txn with index %d", opName, gi)
 		return
 	}
 
-	cid, err := cx.getCreatableID(groupIdx)
+	cid, err := cx.getCreatableID(gi)
```

Comment:
`getCreatableID` is being called only from here, and in this case, the error object is consistently being ignored.
Could you restructure this code so that only a single error would be generated and returned ?

Reply:
The error is not ignored, it gets returned implicitly at the end of opGaidImpl. But I take your point - I can make this clearer. I'll probably eliminate `getCreatableID()`

Reply:
That's what I did. getCreatableID is gone, so the error handling is clearer now.